From 7400aa0f6bb945806c7aca0b07001ef1011e1c79 Mon Sep 17 00:00:00 2001
From: Harshitha Prem <quic_hprem@quicinc.com>
Date: Tue, 7 Mar 2023 23:36:13 +0530
Subject: [PATCH] ath12k: fix tcl_metadata vdev_id info

During AP-STA repeater scenario, initially,
AP interface is brought up with vdev_id 0 and
in parallel, STA interface is brought up with
vdev_id 1 and scan is initiated for STA interface.
While the scan is on going, AP interface is disabled
as the STA is not yet associated. Hence, vdev_id 0 is
deleted.

Once STA scan is completed, vdev (vdev_id 1) is deleted
and during association again STA interface is created with
vdev_id 0. During this, the vdev_id information is appended
in tcl_metadata which results in incorrect configuration of
vdev_id because the ahvif's deflink still had the old vdev_id
in the tcl_metadata information.

Hence, add changes to write tcl_metadata config instead of
appending it.

Signed-off-by: Harshitha Prem <quic_hprem@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1159,11 +1159,11 @@ void ath12k_dp_vdev_tx_attach(struct ath
 {
 	struct ath12k_base *ab = ar->ab;
 
-	arvif->tcl_metadata |= HTT_TCL_META_DATA_GET(1, HTT_TCL_META_DATA_TYPE) |
-			       HTT_TCL_META_DATA_GET(arvif->vdev_id,
-						     HTT_TCL_META_DATA_VDEV_ID) |
-			       HTT_TCL_META_DATA_GET(ar->pdev->pdev_id,
-						     HTT_TCL_META_DATA_PDEV_ID);
+	arvif->tcl_metadata = HTT_TCL_META_DATA_GET(1, HTT_TCL_META_DATA_TYPE) |
+			      HTT_TCL_META_DATA_GET(arvif->vdev_id,
+						    HTT_TCL_META_DATA_VDEV_ID) |
+			      HTT_TCL_META_DATA_GET(ar->pdev->pdev_id,
+						    HTT_TCL_META_DATA_PDEV_ID);
 
 	/* set HTT extension valid bit to 0 by default */
 	arvif->tcl_metadata &= ~HTT_TCL_META_DATA_VALID_HTT;
